### DISCUSSION TITLE
How to change the Build Action of a PhysicalFile?

### DISCUSSION THREAD

**Discussion by GFlisch** [VsixCommunity/Community.VisualStudio.Toolkit#312] [Q&A] [answered]
I am using the toolkit and I add a file to a project.

After this I want to change the Build Action Property from None to Resource.

I have first tried with the TrySetAttributeAsync method but it fails.

I have tried by using the SetProperty method of the IVsHierarchy but it fails.

PhysicalFile menuXml ... code to create the PhysicalFile
menuXml.GetItemInfo(out var hierarchy, out uint itemId, out var hierarchyItem);
var r = hierarchy.SetProperty(itemId, (int)__VSHPROPID4.VSHPROPID_BuildAction, "Resource");

r = -2147352573

Any idea to change the Build Action?

Thanks.

**Comment by madskristensen** [comment]
@reduckted any idea?

**Comment by reduckted** [comment, marked as answer]
I found this answer on Stack Overflow which gives a solution that works (with a few modifications): https://stackoverflow.com/a/24538728/4397397

ℹ️ _See my comment below for an even simpler solution._

Here's a simplified solution:
```cs
private void SetBuildAction(PhysicalFile file, string buildAction)
{
    ThreadHelper.ThrowIfNotOnUIThread();

    file.GetItemInfo(out IVsHierarchy hierarchy, out uint itemId, out _);

    int result = hierarchy.SetProperty(itemId, (int)__VSHPROPID4.VSHPROPID_BuildAction, buildAction);
    if (ErrorHandler.Succeeded(result))
    {
        return;
    }

    result = hierarchy.GetProperty(itemId, (int)__VSHPROPID.VSHPROPID_BrowseObject, out object browseObject);
    if (ErrorHandler.Succeeded(result))
    {
        PropertyDescriptorCollection propertyDescriptors = TypeDescriptor.GetProperties(browseObject);
        if (propertyDescriptors != null)
        {
            PropertyDescriptor itemTypeDescriptor = GetPropertyDescriptor(propertyDescriptors, "ItemType");
            if (itemTypeDescriptor != null)
            {
                try
                {
                    itemTypeDescriptor.SetValue(
                        browseObject,
                        itemTypeDescriptor.Converter.ConvertFromInvariantString(buildAction)
                    );
                    return;
                }
                catch (NotSupportedException)
                {
                    // Suppress this error. We'll throw our
                    // own `NotSupportedException` further below.
                }
            }
        }
    }

    throw new NotSupportedException($"Could not set the build action of '${file.FullPath}'.");
}

private static PropertyDescriptor GetPropertyDescriptor(PropertyDescriptorCollection properties, string name)
{
    // The property names seem to be wrapped in {}, so try to find the given
    // name first, and if that doesn't work, then wrap the name in braces.
    return properties[name] ?? properties[$"{{{name}}}"];
}
```

**Reply by GFlisch** [reply]
It works. Thanks.

**Reply by reduckted** [reply]
It turns out you can actually do this with the toolkit today (the code in the toolkit even refers to the same StackOverflow answer 😄). You just need to know what property descriptor name to use, and that's definitely not obvious.

Using the toolkit:
```cs
await file.TrySetAttributeAsync("{ItemType}", "None");
```

@madskristensen  Do you think this adding a dedicated method to `PhysicalFile` for this would be a good idea? Having to use `"{ItemType}"`  as the property descriptor name isn't very intuitive.

**Reply by madskristensen** [reply]
Yes, adding this makes good sense to me. It will definitely make the toolkit easier and more intuitive to work with for these situations

**Reply by GFlisch** [reply]
Hi @madskristensen,

For me it makes indeed sense that the TrySetAttributeAsync method is used. I use this method to change the Custom Tool or the DependentUpon, Copy to Output Directory, etc... properties.
Now I have tried the {ItemType} and it doesn't work to change a xaml Build Action from Page to Application Definition but the SetBuildAction method works.

To be consistent, on my side, I think that the best is to change the TrySetAttributeAsync implementation and when the name is "Build Action" use the code provided by @reduckted. Doing this it will be clear for the developers than when we have to change a property we just look the property name displayed in the Visual Studio File Properties and simply use it in the method TrySetAttributeAsync.

**Reply by madskristensen** [reply]
That sounds reasonable

**Reply by reduckted** [reply]
I've created #315 with a proposed solution for this.
